dot/state/pruner.go

```diff
@@ -0,0 +1,274 @@
+	blockHash *common.Hash
+	// Hash of keys that are inserted into state trie of the block
+	insertedKeys []*common.Hash
+	// Hash of keys that are deleted from state trie of the block
+	deletedKeys []*common.Hash
+}
+
+func newJournalRecord(hash *common.Hash, insertedKeys, deletedKeys []*common.Hash) *journalRecord {
+	return &journalRecord{
+		blockHash:    hash,
+		insertedKeys: insertedKeys,
+		deletedKeys:  deletedKeys,
+	}
+}
+
+type deathRow struct {
+	blockHash   *common.Hash
+	deletedKeys map[*common.Hash]int64 // keys hash that will be deleted from DB
+}
+
+// CreatePruner creates a pruner
+func CreatePruner(db chaindb.Database, retainBlocks int64) (Pruner, error) {
+	p := &FullNodePruner{
+		deathList:    make([]*deathRow, 0),
+		deathIndex:   make(map[*common.Hash]int64),
+		storageDB:    chaindb.NewTable(db, storagePrefix),
+		journalDB:    chaindb.NewTable(db, journalPrefix),
+		retainBlocks: retainBlocks,
+	}
+
+	blockNum, err := p.getLastPrunedIndex()
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("pruner", "last pruned block", blockNum)
+	blockNum++
+
+	p.pendingNumber = blockNum
+
+	// load deathList and deathIndex from journalRecord
+	for {
+		record, err := p.getJournalRecord(blockNum)
+		if err == chaindb.ErrKeyNotFound {
+			break
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		err = p.addDeathRow(record, blockNum)
+		if err != nil {
+			return nil, err
+		}
+
+		blockNum++
+	}
+
+	go p.start()
+
+	return p, nil
+}
+
+// StoreJournalRecord stores journal record into DB and add deathRow into deathList
+func (p *FullNodePruner) StoreJournalRecord(deleted, inserted []*common.Hash, blockHash *common.Hash, blockNum *big.Int) error {
+	jr := newJournalRecord(blockHash, inserted, deleted)
+
+	err := p.storeJournal(blockNum.Int64(), jr)
+	if err != nil {
+		return fmt.Errorf("failed to store journal record for %d: %w", blockNum, err)
+	}
+
+	logger.Info("journal record stored", "block", blockNum.Int64())
+	err = p.addDeathRow(jr, blockNum.Int64())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *FullNodePruner) addDeathRow(jr *journalRecord, blockNum int64) error {
+	p.Lock()
+	defer p.Unlock()
+
+	// remove re-inserted keys
+	for _, k := range jr.insertedKeys {
+		if num, ok := p.deathIndex[k]; ok {
+			delete(p.deathList[num-p.pendingNumber].deletedKeys, k)
+			delete(p.deathIndex, k)
+		}
+	}
+
+	// add deleted keys from journal to death index
+	for _, k := range jr.deletedKeys {
+		p.deathIndex[k] = blockNum
+	}
+
+	deletedKeys := make(map[*common.Hash]int64)
+	for _, data := range jr.deletedKeys {
+		deletedKeys[data] = blockNum
+	}
+
+	dr := &deathRow{
+		blockHash:   jr.blockHash,
+		deletedKeys: deletedKeys,
+	}
+
+	// add deathRow to deathList
+	p.deathList = append(p.deathList, dr)
+	return nil
+}
+
+func (p *FullNodePruner) start() {
+	logger.Info("pruning started")
+
+	for {
+		p.Lock()
+		if int64(len(p.deathList)) <= p.retainBlocks {
+			p.Unlock()
+			time.Sleep(2 * time.Second)
```

Comment:
why is the sleep here? pls change to constant

Reply:
The infinite loop would otherwise consume a lot of CPU cycles by continuously checking for the condition.